### PR TITLE
Initial setup of unit tests

### PR DESCRIPTION
**Pull request by Achal1607**
This PR introduces an initial setup for unit testing to help validate fixes more quickly, without the need for a full NetBeans environment and VS Code UI runs. This setup includes sample unit tests for `JDKDownloaderView` in `view.ts` and adds few development dependencies for testing and mocking. A few VS Code-related mocks are also included.

### **List of DevDependencies added and Justification:**
  
- **`@istanbuljs/nyc-config-typescript`:**  
  A code coverage tool, with TypeScript support, ensuring code coverage reporting for TypeScript code.

- **`@types/chai` & `chai`:**  
  `chai` is a popular assertion library used for writing readable and expressive test assertions. `@types/chai` provides TypeScript definitions for `chai`.

- **`@types/sinon` & `sinon`:**  
  `sinon` is a  library for creating mocks, spies, and stubs, useful for isolating and testing code components. `@types/sinon` provides TypeScript definitions.

- **`ts-mockito`:**  
  Provides a simpler API for creating mocks in TypeScript, offering an alternative to `sinon` for creating clean, readable mocks in TypeScript. Sinon doesn't have support for mocking interfaces and namesapce. So, it is also required.

- **`vscode-uri`:**  
  Used for handling and manipulating URIs in a way consistent with VS Code’s own URI utilities, especially useful for file and path manipulations within tests.

### CODE REVIEW

**Review by sid-srini — CHANGES_REQUESTED** [review]
Thanks a lot @Achal1607 for getting the unit tests framework added.

Please do rebase against main after merging #310. 
- That should remove many of the files/changes from the scope of this PR.
- Some work maybe needed to incorporate the changes on newly added tests on *main* like the localisation tests.

Most of my comments are queries for clarification, with few minor code suggestions.

Thanks a lot.

File: patches/add-licences.diff
[the remark this answers is not in the document]
Comment (Achal1607) [reply patches/add-licences.diff:1]: Yes all these are dev-dependencies, but currently `PrepareBundles` checks for licenses of all the dependencies during `ant build`. For running test targets it doesn't check for the licenses.
So, do you suggest removing check for licenses of dev-dependencies from the `ant build` target?

File: vscode/src/test/unit/testUtils.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            return "macOS";
+        default:
+            return "windows";
+    }
+}
+
+export const getMachineArch = (): string => {
+    switch (process.arch) {
+        case "arm64":
+            return "aarch64";
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/testUtils.ts:33]: I didn't get it, can you please explain?

File: vscode/src/test/integration/constants.ts
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/integration/constants.ts:1]: Done

File: vscode/src/test/unit/mocks/vscode/uri.ts [hunk 2]
```diff
@@ -0,0 +1,731 @@
+/* eslint-disable max-classes-per-file */
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/mocks/vscode/uri.ts:5]: Done

File: vscode/src/test/unit/mocks/vscode/charCode.ts [hunk 3]
```diff
@@ -0,0 +1,425 @@
+/* eslint-disable camelcase */
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/mocks/vscode/charCode.ts:5]: Done

File: vscode/package.json
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/package.json:1]: Chai v5 is not compatible with our repository, I tried installing it first but it doesn't allow named exports. Reference: https://github.com/chaijs/chai/issues/1561

File: vscode/src/test/unit/mocks/vscode/namespaces/window.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    const mockedOutputChannel = mock<vscode.OutputChannel>();
+    when(mockedWindow.createOutputChannel(
+        anyString()
+    )).thenReturn(instance(mockedOutputChannel));
+} 
+
+const mockMessageView = () => {
+    when(mockedWindow.showErrorMessage(anyString())).thenReturn(Promise.resolve(anyString()));
+    when(mockedWindow.showInformationMessage(anyString())).thenReturn(Promise.resolve(undefined));
+    when(mockedWindow.showWarningMessage(anyString())).thenReturn(Promise.resolve(undefined));
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/mocks/vscode/namespaces/window.ts:51]: Done

File: patches/add-licences.diff
[the remark this answers is not in the document]
Comment (sid-srini) [reply patches/add-licences.diff:1]: Ok. I was only unsure of the origin of where this subset of licenses was coming from.

> So, do you suggest removing check for licenses of dev-dependencies from the `ant build` target?

No, it is fine to keep the check the licenses for dev-dependencies too.
However, we should ensure that the ant task does not include dev-dependency licenses in the the build distribution. I would assume that is out of scope of this PR.

File: vscode/src/test/unit/mocks/vscode/charCode.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sid-srini) [reply vscode/src/test/unit/mocks/vscode/charCode.ts:5]: Sorry, I'm not able to see any change in this file header in the current commits in the PR.

File: vscode/src/test/unit/mocks/vscode/uri.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sid-srini) [reply vscode/src/test/unit/mocks/vscode/uri.ts:5]: Sorry, I'm not able to see any change in this file header in the current commits in the PR.

[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/mocks/vscode/uri.ts]: Sorry, I forgot to push that, I had it on my local. I pushed the changes, can you please take a look at it now.

File: vscode/src/test/unit/mocks/vscode/charCode.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/mocks/vscode/charCode.ts]: Sorry, I forgot to push that, I had it on my local. I pushed the changes, can you please take a look at it now.

File: vscode/src/test/unit/testUtils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sid-srini) [reply vscode/src/test/unit/testUtils.ts:33]: With the changes incorporated in #294, the windows OS download would always return the host arch as "x64", even on an "arm64" machine. The test would perhaps fail if "aarch64" is received for this combination.

[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/test/unit/testUtils.ts:33]: Done

File: patches/add-licences.diff [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
++
++  If the disclaimer of warranty and limitation of liability provided
++above cannot be given local legal effect according to their terms,
++reviewing courts shall apply local law that most closely approximates
++an absolute waiver of all civil liability in connection with the
++Program, unless a warranty or assumption of liability accompanies a
++copy of the Program in return for a fee.
++
++                     END OF TERMS AND CONDITIONS
++
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply patches/add-licences.diff]: Done

File: build.xml [hunk 6]
```diff
@@ -62,7 +62,7 @@
             patches/rename-debugger.diff
             patches/remove-db.diff
             patches/nbjavac-not-required.diff
-            patches/l10n-licence.diff
+            patches/add-licences.diff
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply build.xml]: Separated  _l10n-licence.diff_ and _dev-dependencies-licenses.diff_. Thanks

File: vscode/package.json
[the remark this answers is not in the document]
Comment (sid-srini) [reply vscode/package.json:1]: Ok. Thanks.
Should we then use the latest in the previous major version?
- chai: 4.5.0

[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/package.json:1]: I upgraded it to 4.5.0, it is working fine.

**Review by sid-srini — APPROVED** [review]
Thanks a lot @Achal1607. LGTM 👍

**Pull request merged**